### PR TITLE
6 ui entity screens

### PR DESCRIPTION
**Pull request by boltic28**

### CODE REVIEW

File: app/src/main/java/com/boltic28/taskmanager/businesslayer/CaseGoalStructureImpl.kt [hunk 1]
```diff
@@ -1,61 +1,62 @@
...
+    private val goalRepository: GoalRepository
+) : CaseGoalStructure {
 
     override fun setChildrenFor(goal: Goal): Single<Goal> =
         Single.just(goal)
-            .zipWith(keyRepository.getAllForGoal(goal.id),
-                BiFunction<Goal, List<KeyResult>, Goal> { sGoal, keys ->
-                    sGoal.copy(keys = keys)
-                })
+            .zipWith(keyRepository.getAllForGoal(goal.id), { sGoal, keys ->
```
Comment (AlexeyZubkovskiyITechArt) [inline app/src/main/java/com/boltic28/taskmanager/businesslayer/CaseGoalStructureImpl.kt]: Combine all values in one method instead copying, zip all values in one place

File: app/src/main/java/com/boltic28/taskmanager/businesslayer/CaseGoalStructureImpl.kt [hunk 2]
```diff
@@ -1,61 +1,62 @@
...
 import com.boltic28.taskmanager.datalayer.room.idea.IdeaRepository
 import com.boltic28.taskmanager.datalayer.room.keyresult.KeyRepository
 import com.boltic28.taskmanager.datalayer.room.step.StepRepository
 import com.boltic28.taskmanager.datalayer.room.task.TaskRepository
-import io.reactivex.Observable
 import io.reactivex.Single
-import io.reactivex.functions.BiFunction
+import io.reactivex.schedulers.Schedulers
 
 class CaseGoalStructureImpl(
```
Comment (AlexeyZubkovskiyITechArt) [inline app/src/main/java/com/boltic28/taskmanager/businesslayer/CaseGoalStructureImpl.kt]: Change naming

File: app/src/main/java/com/boltic28/taskmanager/businesslayer/IdeaFragmentInteractor.kt [hunk 3]
```diff
@@ -14,11 +14,13 @@ interface IdeaFragmentInteractor {
 
     fun getIdeaById(id: Long): Single<Idea>
 
-    fun getAllElements(): Single<List<Any>>
+    fun getFreeStepsGoalsKeys(): Single<List<Any>>
```
Comment (AlexeyZubkovskiyITechArt) [inline app/src/main/java/com/boltic28/taskmanager/businesslayer/IdeaFragmentInteractor.kt]: Why do you need the list of Any?

File: app/src/main/java/com/boltic28/taskmanager/businesslayer/IdeaFragmentInteractorImpl.kt [hunk 4]
```diff
@@ -50,11 +50,16 @@ class IdeaFragmentInteractorImpl(
...
+    override fun create(item: Step): Single<Long> =
+        stepRepository.insert(item)
+
+    override fun create(item: Goal): Single<Long> =
+        goalRepository.insert(item)
 
-    override fun create(item: Goal): Single<Long> {
-        TODO("Not yet implemented")
+    override fun getParentName(item: Idea): Single<String> {
+            if (item.goalId != 0L) return goalRepository.getById(item.goalId).map { it.name }
```
Comment (AlexeyZubkovskiyITechArt) [inline app/src/main/java/com/boltic28/taskmanager/businesslayer/IdeaFragmentInteractorImpl.kt]: 0L should be constant in this case

File: app/src/main/java/com/boltic28/taskmanager/businesslayer/IdeaFragmentInteractorImpl.kt [hunk 5]
```diff
@@ -50,11 +50,16 @@ class IdeaFragmentInteractorImpl(
...
+            if (item.stepId != 0L) return stepRepository.getById(item.stepId).map { it.name }
+            if (item.keyId != 0L) return keyRepository.getById(item.keyId).map { it.name }
+            return Single.just("error")
```
Comment (AlexeyZubkovskiyITechArt) [inline app/src/main/java/com/boltic28/taskmanager/businesslayer/IdeaFragmentInteractorImpl.kt]: It is bad idea to send error in conditions

File: app/src/main/java/com/boltic28/taskmanager/businesslayer/StepFragmentInteractorImpl.kt [hunk 6]
```diff
@@ -52,31 +52,31 @@ class StepFragmentInteractorImpl(
...
         } else {
-            when ((done * 100.0 )/ pointsToFull) {
-                100.0 -> step.copy(progress = Progress.DONE)
-                in 75.0..99.0 -> step.copy(progress = Progress.PROGRESS_80)
-                in 55.0..74.0 -> step.copy(progress = Progress.PROGRESS_60)
-                in 35.0..54.0 -> step.copy(progress = Progress.PROGRESS_40)
-                in 15.0..34.0 -> step.copy(progress = Progress.PROGRESS_20)
-                else -> step.copy(progress = Progress.PROGRESS_0)
+            when ((done * 100.0) / pointsToFull) {
+                100.0 -> step.copy(progress = Progress.DONE, isDone = true, isStarted = isStarted)
```
Comment (AlexeyZubkovskiyITechArt) [inline app/src/main/java/com/boltic28/taskmanager/businesslayer/StepFragmentInteractorImpl.kt]: Duplication of logic.

File: app/src/main/java/com/boltic28/taskmanager/ui/adapter/controllers/HolderController.kt [hunk 7]
```diff
@@ -37,7 +38,22 @@ abstract class HolderController {
         holder.type == getType()
 
     interface OnActionClickListener {
+        fun isNeedToShowConnection(): Boolean
         fun onActionButtonClick(item: Any)
         fun onViewClick(item: Any)
     }
+
+    fun fetchDescription(description: String): String =
+        if (description.length > 78) {
```
Comment (AlexeyZubkovskiyITechArt) [inline app/src/main/java/com/boltic28/taskmanager/ui/adapter/controllers/HolderController.kt]: numbers should be named

File: app/src/main/java/com/boltic28/taskmanager/businesslayer/TaskFragmentInteractorImpl.kt [hunk 8]
```diff
@@ -16,4 +26,32 @@ class TaskFragmentInteractorImpl(
 
     override fun getTaskById(id: Long): Single<Task> =
         taskRepository.getById(id)
+
+    override fun getParentName(item: Task): Single<String> {
+        if (item.goalId != 0L) return goalRepository.getById(item.goalId).map { it.name }
```
Comment (AlexeyZubkovskiyITechArt) [inline app/src/main/java/com/boltic28/taskmanager/businesslayer/TaskFragmentInteractorImpl.kt:35]: duplication of logic

File: app/src/main/java/com/boltic28/taskmanager/ui/adapter/controllers/HolderController.kt [hunk 9]
```diff
@@ -37,7 +38,22 @@ abstract class HolderController {
...
+            description.substring(0..75) + "..."
+        } else {
+            description
+        }
+
+    fun fetchName(name: String): String =
+        if (name.length > 23) {
```
Comment (AlexeyZubkovskiyITechArt) [inline app/src/main/java/com/boltic28/taskmanager/ui/adapter/controllers/HolderController.kt]: Beter to move this to extenesion and pass quantity as argument

File: app/src/main/java/com/boltic28/taskmanager/ui/adapter/controllers/StepSmallViewController.kt [hunk 10]
```diff
@@ -22,23 +22,27 @@ class StepSmallViewController : HolderController() {
...
-        if (item.goalId == 0L && item.keyId == 0L){
-            button.setImageResource(R.drawable.ic_link)
+        val status: ImageView = itemView.findViewById(R.id.small_step_image_status)
+
+        if (item.isStarted) status.setImageResource(R.drawable.ic_started)
+        if (item.progress == Progress.DONE) status.setImageResource(R.drawable.ic_done)
+        name.text = fetchName(item.name)
+        icon.setImageResource(R.drawable.step_ph)
+
+        if (listener.isNeedToShowConnection()) {
```
Comment (AlexeyZubkovskiyITechArt) [inline app/src/main/java/com/boltic28/taskmanager/ui/adapter/controllers/StepSmallViewController.kt]: Use when instead

File: app/src/main/java/com/boltic28/taskmanager/ui/adapter/controllers/StepViewController.kt [hunk 11]
```diff
@@ -25,14 +27,60 @@ class StepViewController: HolderController() {
...
 
-        // describe buttons to other entity
-        // TODO
-        itemView.findViewById<Button>(R.id.step_button_start).setOnClickListener {  }
-        itemView.findViewById<Button>(R.id.step_button_done).setOnClickListener {  }
+        val statusIsAttached: ImageView = itemView.findViewById(R.id.item_step_status_linked)
+        val statusAttention: ImageView = itemView.findViewById(R.id.item_step_status_attention)
+        val statusStarted: ImageView = itemView.findViewById(R.id.item_step_status_started)
+        val statusDone: ImageView = itemView.findViewById(R.id.item_step_status_done)
+        val days = TimeUnit.DAYS.toDays(
```
Comment (AlexeyZubkovskiyITechArt) [inline app/src/main/java/com/boltic28/taskmanager/ui/adapter/controllers/StepViewController.kt]: to method

File: app/src/main/java/com/boltic28/taskmanager/ui/adapter/controllers/TaskViewController.kt [hunk 12]
```diff
@@ -24,15 +27,65 @@ class TaskViewController: HolderController() {
...
+        val statusStarted: ImageView = itemView.findViewById(R.id.item_task_status_started)
+        val statusDone: ImageView = itemView.findViewById(R.id.item_task_status_done)
+        val days = TimeUnit.DAYS.toDays(
+            LocalDateTime.now().toEpochSecond(ZoneOffset.UTC) - item.dateClose.toEpochSecond(
+                ZoneOffset.UTC
+            )
+        )
+
+        statusCycling.setColorFilter(
+            if (item.cycle == Cycle.NOT_CYCLE)
```
Comment (AlexeyZubkovskiyITechArt) [inline app/src/main/java/com/boltic28/taskmanager/ui/adapter/controllers/TaskViewController.kt]: Better use braces everytime except situations when you can write if-else in one line

File: app/src/main/java/com/boltic28/taskmanager/ui/screens/ideafragment/IdeaFragment.kt [hunk 13]
```diff
@@ -11,48 +11,66 @@ import com.boltic28.taskmanager.ui.screens.mainfragment.MainFragment.Companion.I
...
+        idea_fr_button_home.setOnClickListener {
+            findNavController().navigate(R.id.mainFragment)
+        }
+    }
+
+    private fun setButtonOwner(idea: Idea) {
+        idea_fr_recycler.visibility = View.INVISIBLE
+        idea_fr_its_elements.visibility = View.INVISIBLE
+        if (idea.goalId != 0L || idea.keyId != 0L || idea.stepId != 0L) {
+            model.disposables + model.interactor.getParentName(idea)
```
Comment (AlexeyZubkovskiyITechArt) [inline app/src/main/java/com/boltic28/taskmanager/ui/screens/ideafragment/IdeaFragment.kt]: View mustn't access interactor directly

File: app/src/main/java/com/boltic28/taskmanager/businesslayer/CaseGoalStructureImpl.kt (hunk 1 above)
Comment (boltic28) [reply app/src/main/java/com/boltic28/taskmanager/businesslayer/CaseGoalStructureImpl.kt]: done

File: app/src/main/java/com/boltic28/taskmanager/businesslayer/CaseGoalStructureImpl.kt (hunk 2 above)
Comment (boltic28) [reply app/src/main/java/com/boltic28/taskmanager/businesslayer/CaseGoalStructureImpl.kt]: renamed to ItemStructureProvider
it works with all items

File: app/src/main/java/com/boltic28/taskmanager/businesslayer/IdeaFragmentInteractorImpl.kt (hunk 4 above)
Comment (boltic28) [reply app/src/main/java/com/boltic28/taskmanager/businesslayer/IdeaFragmentInteractorImpl.kt]: done

File: app/src/main/java/com/boltic28/taskmanager/businesslayer/IdeaFragmentInteractorImpl.kt (hunk 5 above)
Comment (boltic28) [reply app/src/main/java/com/boltic28/taskmanager/businesslayer/IdeaFragmentInteractorImpl.kt]: corrected. It sends name of item if can not find a name of parent

**Pull request merged**